### ISSUE TITLE
Proposal: Append EDNS0 Data for Proxy

### ISSUE THREAD

**Issue by johnbelamaric** [coredns/coredns#468]
I have a use case where I want to configure CoreDNS to append a custom EDNS0 option before sending data out via the proxy. For now, the EDNS0 option will be static for any given CoreDNS proxy, so it can be explicitly specified in the Corefile.

Should this be its own middleware? Or part of rewrite?

Thoughts?

**Comment by miekg** [comment]
[ Quoting <notifications@github.com> in "[miekg/coredns] Proposal: Append ES..." ]
>I have a use case where I want to configure CoreDNS to append a custom EDNS0 option before sending data out via the proxy. For now, the EDNS0 option will be static for any given CoreDNS proxy, so it can be explicitly specified in the Corefile.
>
>Should this be its own middleware? Or part of rewrite?
>
>Thoughts?

Look at the Caddy proxy middleware, you could add headers there. I think we
should allow the same in our middleware. I ripped that stuff out, because it was
too hard to think about it at the time. So yes, want! But the syntax and how and
what may be slightly header to specify.

The rewrite middleware is not the place (I think)....

**Comment by johnbelamaric** [comment]
That would work. But if it's only in the proxy middleware, you couldn't use it to add options to the query for use by downstream middleware.

**Comment by miekg** [comment]
[ Quoting <notifications@github.com> in "Re: [miekg/coredns] Proposal: Appen..." ]
>That would work. But if it's only in the proxy middleware, you couldn't use it to add options to the query for use by downstream middleware.

that is true, maybe *rewrite* is the better place. At least there it is more
generic. Be aware that the *rewrite* middleware still needs a lot of love.

**Comment by johnbelamaric** [comment]
What about reworking the rewrite to work something like this:

```
rewrite <field> <from> <to>
```

Here `<field>` is ”class” ”qtype” ”qname” ”edns0” (and maybe ”sourceip”, ”header” and some others), and ”from condition” can be something like these examples:

```
rewrite qtype ANY A
rewrite qname foo.com bar.com
rewrite class CH IN
rewrite edns0 {code: 0xffee, data: 0x1234} {data: 0x4321} # rewrite data if code and data match
rewrite edns0 {code: 0xffee, data: 0x1234} {code: 0xffea, data: 0x4321} # rewrite code and data if code and data match
rewrite edns0 {code: 0xffee} {code: 0xffea, data: 0x4321} # rewrite code and data if code match
```

I think code would be required in the edns0 matches above. If the edns0 data above doesn't exist, I am thinking it would do nothing. If instead you want to append we could does something like:

```
rewrite edns0 append {code: 0xffee, data: 0x1234}
```

If you wanted to rewrite if it exists and append otherwise:

```
rewrite edns0 replace {code: 0xffee} {code: 0xffee, data: 0x1234}
```

We can allow use of the EDNS0 data structures defined in https://github.com/miekg/dns like this:

```
rewrite edns0 append {code: EDNS0SUBNET, family: 1, sourcenetmask: 32, sourcescope: 0, address: 10.1.1.1}
```

**Comment by johnbelamaric** [comment]
Tag @bwaz since he will probably be the one to implement.

**Comment by miekg** [comment]
Yes, this makes sense. If think we should just go ahead and implement and keep in the back of our minds that it may change in the future (I still like a comprehense DSL in coredns for rewrites/acl and other filtering).  See #469 

Regarding your last example, for `address` somebody will undoubtedly ask: can this be dynamically discovered by CoreDNS, or you may want to put in interface there. etc. etc.

We just might need to flesh this out a little bit more feature wise (and maybe bikeshed a bit on the syntax as well).

**Comment by bwaz** [comment]
https://github.com/miekg/coredns/pull/497

**Comment by johnbelamaric** [comment]
Implemented with #561

**Comment by miekg** [comment]
Should update the release wiki for 007 with this (just started that doc)
